object-store/src/main/java/org/hypertrace/config/objectstore/ConfigObjectScopeResolver.java

```diff
@@ -0,0 +1,131 @@
+package org.hypertrace.config.objectstore;
+
+import com.google.protobuf.Descriptors;
+import com.google.protobuf.Message;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+/**
+ * Resolver class to resolve configs based on scope
+ *
+ * @param <T> Raw config type
+ * @param <U> Resolved config type
+ * @param <S> Scope
+ */
+public abstract class ConfigObjectScopeResolver<T extends Message, U extends Message, S> {
+
+  protected final Optional<U> defaultConfig;
+
+  public ConfigObjectScopeResolver(U defaultConfig) {
+    this.defaultConfig = Optional.of(defaultConfig);
+  }
+
+  public ConfigObjectScopeResolver() {
+    this.defaultConfig = Optional.empty();
+  }
+
+  public List<U> getAllResolvedConfigData(List<T> configDataList) {
+    Map<S, T> configDataMap = getConfigDataMap(configDataList);
+    return configDataList.stream()
+        .flatMap(configData -> resolveConfig(extractScope(configData), configDataMap).stream())
+        .collect(Collectors.toUnmodifiableList());
+  }
+
+  public Optional<U> getResolvedData(List<T> configDataList, S scope) {
+    Map<S, T> configDataMap = getConfigDataMap(configDataList);
+    return resolveConfig(scope, configDataMap);
+  }
+
+  /**
+   * Method to extract scope object from given config object
+   *
+   * @param configData
+   * @return Scope object
+   */
+  protected abstract S extractScope(T configData);
+
+  /**
+   * Method to return scopes with which the given scope needs to be resolved in increasing order of
+   * priority For example, API scope would return a list {customer, env, service, api}
+   *
+   * @param scopeObject
+   * @return list of associated scopes in increasing order of priority
+   */
+  protected abstract List<S> getResolutionScopesWithIncreasingPriority(S scopeObject);
+
+  /**
+   * Method to return object of resolved config type from object store config data
+   *
+   * @param configStoreData
+   * @return
+   */
+  protected abstract U convertConfig(T configStoreData);
+
+  /**
+   * Method to resolve config store object with object of resolved config type
+   *
+   * @param fallbackConfig
+   * @param preferredConfig
+   * @return
+   */
+  protected abstract U resolveConfig(U fallbackConfig, T preferredConfig);
+
+  private Map<S, T> getConfigDataMap(List<T> configDataList) {
+    return configDataList.stream()
+        .collect(Collectors.toMap(this::extractScope, Function.identity()));
+  }
+
+  private Optional<U> resolveConfig(S scope, Map<S, T> configDataMap) {
+    Optional<T> mergedConfig =
+        getResolutionScopesWithIncreasingPriority(scope).stream()
+            .flatMap(scopeObject -> Optional.ofNullable(configDataMap.get(scope)).stream())
+            .reduce(
+                (fallbackConfig, preferredConfig) ->
+                    (T) mergeConfigs(fallbackConfig, preferredConfig));
```

Comment:
both inputs are T right? we should fix the typing on the method such that both args and the response are same type without a cast. Something like
```java
 private <M extends Message> M mergeConfigs(M fallbackConfig, M preferredConfig) {
```